persistence/sql/persister_recovery.go

```diff
@@ -129,66 +131,127 @@ func (p *Persister) DeleteExpiredRecoveryFlows(ctx context.Context, expiresAt ti
 	return nil
 }
 
-func (p *Persister) CreateRecoveryCode(ctx context.Context, recoveryCode *code.RecoveryCode) error {
+func (p *Persister) CreateRecoveryCode(ctx context.Context, dto *code.RecoveryCodeDTO) (*code.RecoveryCode, error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateRecoveryCode")
 	defer span.End()
 
-	code := recoveryCode.Code
-	recoveryCode.Code = p.hmacValue(ctx, code)
-	recoveryCode.NID = p.NetworkID(ctx)
+	now := time.Now()
+
+	recoveryCode := &code.RecoveryCode{
+		ID:         x.NewUUID(),
+		Code:       p.hmacValue(ctx, dto.Code),
+		ExpiresAt:  now.UTC().Add(dto.ExpiresIn),
+		IssuedAt:   now,
+		CodeType:   dto.CodeType,
+		FlowID:     dto.FlowID,
+		NID:        p.NetworkID(ctx),
+		IdentityID: dto.IdentityID,
+	}
+
+	if dto.RecoveryAddress != nil {
+		recoveryCode.RecoveryAddress = dto.RecoveryAddress
+		recoveryCode.RecoveryAddressID = uuid.NullUUID{
+			UUID:  dto.RecoveryAddress.ID,
+			Valid: true,
+		}
+	}
 
 	// This should not create the request eagerly because otherwise we might accidentally create an address that isn't
 	// supposed to be in the database.
 	if err := p.GetConnection(ctx).Create(recoveryCode); err != nil {
-		return err
+		return nil, err
 	}
 
-	recoveryCode.Code = code
-	return nil
+	return recoveryCode, nil
 }
 
 func (p *Persister) UseRecoveryCode(ctx context.Context, fID uuid.UUID, codeVal string) (*code.RecoveryCode, error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UseRecoveryCode")
 	defer span.End()
 
-	var recoveryCode code.RecoveryCode
+	var recoveryCode *code.RecoveryCode
 
 	nid := p.NetworkID(ctx)
+
+	flowTableName := new(recovery.Flow).TableName(ctx)
+
 	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
+
+		/* #nosec G201 TableName is static */
+		if err = sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("UPDATE %s SET submit_count = submit_count + 1 WHERE id = ? AND nid = ?", flowTableName), fID, nid).Exec()); err != nil {
+			return
+		}
+
+		var submitCount int
+		// Because MySQL does not support "RETURNING" clauses, but we need the updated `submit_count` later on.
+		/* #nosec G201 TableName is static */
+		if err := sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("SELECT submit_count FROM %s WHERE id = ? AND nid = ?", flowTableName), fID, nid).First(&submitCount)); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+
+			return err
+		}
+
+		if submitCount > 5 {
```

Comment:
Handling this here IMO breaks the isolation of the persister, because now we are performing business logic within something responsible of fecthing / writing to the datastore. For the sake of shipping this, I would be OK to keep it as is if we add a big warning to the function godoc.

If you were to extract this logic, I think we would need some method in the persister such as: `IncreaseSubmitCount() (count int, err error)` and use that to perform our business logic in some other service / place.

A good example for this is the session persister, and session manager:

* session persister: writes and reads from the data store
* session manager: does things like setting cookies, changing AAL of a session, ...

Reply:
Yes, I agree. 

I'd like to revisit this, when (if) we are refactoring the recovery flow or when I add verification to the OTP strategy, I have a few ideas to improve this whole data layout a bit more.

Reply:
Ok, could you please add a codedoc for the function, a todo entry, and an issue for it with you assigned? :)

Reply:
Done.

https://github.com/ory/kratos/issues/2785